core/src/main/scala/cats/collections/TreeList.scala

```diff
@@ -0,0 +1,867 @@
+              a
+            case Balanced(a, l, r) =>
+              // switch the order
+              nexts = r :: l :: Root(a) :: nexts.tail
+              next()
+          }
+    }
+  }
+
+  import Impl._
+
+  private final case class Trees[A](treeList: List[Tree[Nat, A]]) extends TreeList[A] {
+    def prepend[A1 >: A](a1: A1): TreeList[A1] =
+      treeList match {
+        case h1 :: h2 :: rest =>
+          def go[N1 <: Nat, N2 <: Nat, A2 <: A](t1: Tree[N1, A2], t2: Tree[N2, A2]): TreeList[A1] =
+            NatEq.maybeEq[N1, N2](t1.depth, t2.depth) match {
+              case Some(eqv) =>
+                type T[N <: Nat] = Tree[N, A2]
+                Trees(Balanced[N2, A1](a1, eqv.subst[T](t1), t2) :: rest)
+              case None =>
+                Trees(Root(a1) :: treeList)
+            }
+          go(h1, h2)
+        case lessThan2 => Trees(Root(a1) :: lessThan2)
+      }
+    def uncons: Option[(A, TreeList[A])] =
+      treeList match {
+        case Nil => None
+        case Root(a) :: rest => Some((a, Trees(rest)))
+        case Balanced(a, l, r) :: rest => Some((a, Trees(l :: r :: rest)))
+      }
+
+    def get(idx: Long): Option[A] = {
+      @tailrec
+      def loop(idx: Long, treeList: List[Tree[Nat, A]]): Option[A] =
+        if (idx < 0L) None
+        else
+          treeList match {
+            case Nil => None
+            case h :: tail =>
+              if (h.size <= idx) loop(idx - h.size, tail)
+              else h.get(idx)
+          }
+
+      loop(idx, treeList)
+    }
+
+    def lastOption: Option[A] = {
+      @tailrec
+      def loop(treeList: List[Tree[Nat, A]]): Option[A] =
+        treeList match {
+          case Nil => None
+          case head :: tail =>
+            if (tail.isEmpty)
+              head match {
+                case Root(a) => Some(a)
+                case Balanced(_, _, r) => loop(r :: Nil)
+              }
+            else loop(tail)
+        }
+      loop(treeList)
+    }
+
+    def size: Long = {
+      @tailrec
+      def loop(treeList: List[Tree[Nat, A]], acc: Long): Long =
+        treeList match {
+          case Nil => acc
+          case h :: tail => loop(tail, acc + h.size)
+        }
+      loop(treeList, 0L)
+    }
+
+    def foldLeft[B](init: B)(fn: (B, A) => B): B = {
+      @tailrec
+      def loop(init: B, rest: List[Tree[Nat, A]]): B =
+        rest match {
+          case Nil => init
+          case Root(a) :: tail => loop(fn(init, a), tail)
+          case Balanced(a, l, r) :: rest => loop(fn(init, a), l :: r :: rest)
+        }
+
+      loop(init, treeList)
+    }
+
+    def strictFoldRight[B](fin: B)(fn: (A, B) => B): B =
+      treeList.reverse.foldLeft(fin) { (b, treea) =>
+        treea.foldRight(b)(fn)
+      }
+
+    def isEmpty: Boolean = treeList.isEmpty
+    def nonEmpty: Boolean = treeList.nonEmpty
+
+    def map[B](fn: A => B): TreeList[B] = Trees(treeList.map(_.map(fn)))
+
+    def drop(n: Long): TreeList[A] = {
+      @tailrec
```

Comment:
another micro-optimization is to check if n > size and return empty
benchmarks probably can reveal if it's worth it or not